File: kotlin/advent/src/main/kotlin/org/gristle/pdxpuzzles/advent/y2017/Y17D25.kt
package org.gristle.pdxpuzzles.advent.y2017

import org.gristle.pdxpuzzles.advent.utilities.Day
import org.gristle.pdxpuzzles.utilities.parsing.getInts

class Y17D25(private val input: String) : Day {

    enum class Dir {
        LEFT,
        RIGHT,
    }
    data class Action(val write: Boolean, val dir: Dir, val change: Int)
    data class State(val zero: Action, val one: Action)

    override fun part1(): Int {
        val stanzas = input.splitToSequence("\n\n").iterator()
        val steps = stanzas.next()
            .dropLastWhile { c -> !c.isDigit() }
            .takeLastWhile { c -> c.isDigit() }
            .toInt()
        val states = stanzas.asSequence()
            .map { stanza ->
                val args = stanza.lines().map { line ->
                    line.dropLast(1).takeLastWhile { c -> c.isLetterOrDigit() }
                }
                val write0 = args[2] == "1"
                val dir0 = if (args[3] == "left") Dir.LEFT else Dir.RIGHT
                val change0 = args[4][0] - 'A'
                val write1 = args[6] == "1"
                val dir1 = if (args[7] == "left") Dir.LEFT else Dir.RIGHT
                val change1 = args[8][0] - 'A'
                State(
                    Action(write0, dir0, change0),
                    Action(write1, dir1, change1),
                )
            }.toList()
        val slots = ArrayDeque<Boolean>()
        slots.add(false)
        var state = states[0]
        var node = 0

        for (_i in 0 until steps) {
            val action = if (slots[node]) state.one else state.zero
            slots[node] = action.write
            when (action.dir) {
                Dir.LEFT -> if (node == 0) slots.addFirst(false) else node--
                Dir.RIGHT -> {
                    if (node == slots.size - 1) slots.addLast(false)
                    node++
                }
            }
            state = states[action.change]
        }
        return slots.count { it }
    }

    override fun part2() = true
}

fun main() = Day.runDay(Y17D25::class)

//    Class creation: 1ms
//    Part 1: 3745 (76ms)
//    Total time: 77ms

File: rust/advent/src/bin/2017/y17d25.rs
use std::collections::VecDeque;

use itertools::Itertools;
use advent::utilities::get_input::get_input;
use utilities::parsing::get_numbers::ContainsNumbers;
use utilities::structs::stopwatch::{ReportDuration, Stopwatch};

type Input<'a> = &'a str;
type Output = usize;

fn main() {
    let mut stopwatch = Stopwatch::new();
    stopwatch.start();
    let input = get_input(17, 25).unwrap();
    println!("Input parsed ({})", stopwatch.lap().report());
    println!("1. {} ({})", part1(&input), stopwatch.lap().report());
    println!("Total: {}", stopwatch.stop().report());
}

#[derive(Debug)]
enum Dir { Left, Right }

#[derive(Debug)]
struct Action {
    write: bool,
    dir: Dir,
    change: usize,
}

#[derive(Debug)]
struct State {
    zero: Action,
    one: Action,
}

fn penultimate(s: &str) -> &str {
    let ultimate = s.split(' ').rev().next().unwrap();
    &ultimate[..ultimate.len() - 1]
}
fn part1(input: Input) -> Output {
    let (init, states) = input.split_once("\n\n").unwrap();
    let mut init_lines = init.lines();
    let state_idx = penultimate(init_lines.next().unwrap())
        .chars()
        .next()
        .map(|c| (c as u8 - b'A') as usize)
        .unwrap();
    let steps: usize = init_lines.next().unwrap().get_numbers().next().unwrap();

    let states = states.split("\n\n")
        .map(|stanza| {
            let (_, _, write0, dir0, change0, _, write1, dir1, change1) = stanza.lines()
                .map(|line| penultimate(line))
                .collect_tuple()
                .unwrap();
            let write0 = write0 == "1";
            let dir0 = if dir0 == "left" { Dir::Left } else { Dir::Right };
            let change0 = (change0.as_bytes()[0] - b'A') as usize;
            let write1 = write1 == "1";
            let dir1 = if dir1 == "left" { Dir::Left } else { Dir::Right };
            let change1 = (change1.as_bytes()[0] - b'A') as usize;
            State {
                zero: Action { write: write0, dir: dir0, change: change0 },
                one: Action { write: write1, dir: dir1, change: change1 },
            }
        })
        .collect_vec();

    let mut slots = VecDeque::with_capacity(3746);
    slots.push_front(false);
    let mut state = &states[state_idx];
    let mut node = 0;

    for _ in 0..steps {
        let node_val = slots.get_mut(node).unwrap();
        let action = if *node_val { &state.one } else { &state.zero };
        *node_val = action.write;
        match action.dir {
            Dir::Left => {
                if node == 0 {
                    slots.push_front(false);
                } else {
                    node -= 1;
                }
            }
            Dir::Right => {
                if node == slots.len() - 1 {
                    slots.push_back(false);
                }
                node += 1;
            }
        }

        state = &states[action.change];
    }
    slots.into_iter().filter(|&b| b).count()
}
#[test]
fn default() {
    let input = get_input(17, 25).unwrap();
    assert_eq!(3745, part1(&input));
}

// Input parsed (16μs)
// 1. 3745 (21ms)
// Total: 21ms
